src/style-spec/function/evaluation_context.js

```diff
@@ -99,72 +95,103 @@ module.exports = () => ({
         return toString(typeOf(x));
     },
 
-    as: function (value: Value, expectedType: Type, name?: string) {
-        assert(isValue(value), `Invalid value ${JSON.stringify(value)}`);
-        assert(expectedType.kind, `Invalid type ${JSON.stringify(expectedType)}`);
-
-        let type;
-        let typeError = false;
-        if (expectedType.kind === 'Null') {
-            typeError = value === null;
-        } else if (expectedType.kind === 'Value') {
-            typeError = false;
-        } else if (expectedType.kind !== 'Array' && expectedType.kind !== 'Color' && expectedType.kind !== 'Error') {
-            typeError = typeof value !== jsTypes[expectedType.kind];
-        } else {
-            type = typeOf(value);
-            typeError = checkSubtype(expectedType, type);
+    asJSType: function (expectedType: string, values: Array<Function>) {
+        let value;
+        for (let i = 0; i < values.length; i++) {
+            value = values[i]();
+            if (typeof value === expectedType && value !== null) {
+                return value;
+            }
         }
+        const expected = jsTypes[expectedType].kind;
+        throw new RuntimeError(`Expected value to be of type ${expected}, but found ${this.typeOf(value)} instead.`);
+    },
+
+    asArray: function (value: Value, expectedType: ArrayType) {
+        const type = typeOf(value);
+        const typeError = checkSubtype(expectedType, type);
 
         if (typeError) {
-            if (!type) type = typeOf(value);
-            throw new RuntimeError(`Expected ${name || 'value'} to be of type ${toString(expectedType)}, but found ${toString(type)} instead.`);
+            throw new RuntimeError(`Expected value to be of type ${toString(expectedType)}, but found ${toString(type)} instead.`);
         }
 
         return value;
     },
 
-    toColor: function (input: Value) {
-        if (typeof input === 'string') {
-            return this.parseColor(input);
-        } else if (Array.isArray(input) && (input.length === 3 || input.length === 4)) {
-            return this.rgba(...input);
-        } else {
-            throw new RuntimeError(`Could not parse color from value '${JSON.stringify(input)}'`);
+    toColor: function (args: Array<()=>any>) {
+        let input;
+        let error;
+        for (let i = 0; i < args.length; i++) {
+            input = args[i]();
+            error = null;
+            if (typeof input === 'string') {
+                const c = this._parseColor(input);
+                if (c) return c;
+            } else if (Array.isArray(input)) {
+                error = this._validateRGBA(input[0], input[1], input[2], input[3]);
+                if (!error) return new Color(input[0] / 255, input[1] / 255, input[2] / 255, input[3]);
```

Comment:
Can you replace those 2 lines with a call to `rgba`?

Reply:
No, because `rgba` throws a `RuntimeError` on failure.

Reply:
Oh, nevermind — I just realized `toColor` now acts like coalesce through multiple args...